sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotUseCollectionInItsOwnMethodCalls.cs

```diff
@@ -0,0 +1,182 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Helpers.CSharp;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotUseCollectionInItsOwnMethodCalls : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S2114";
+        private const string MessageFormat = "Change one instance of '{0}' to a different value; {1}";
+        private const string AlwaysEmptyCollectionMessage = "This operation always produces an empty collection.";
+        private const string AlwaysSameCollectionMessage = "This operation always produces the same collection.";
+        private const string AlwaysTrueMessage = "Comparing to itself always returns true.";
+        private const string UnexpectedBehaviorMessage = "This operation will probably result in an unexpected behavior.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<string> trackedMethodNames = ImmutableHashSet.Create("AddRange", "Concat",
+            "Union", "Except", "Intersect", "SequenceEqual", "UnionWith", "ExceptWith");
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var invocation = (InvocationExpressionSyntax)c.Node;
+                    var invocationExpressionString = invocation.Expression.ToString();
+
+                    if (!trackedMethodNames.Any(method => invocationExpressionString.EndsWith(method,
+                        StringComparison.Ordinal)))
+                    {
+                        return;
+                    }
+
+                    var operands = GetOperandsToCheckIfTrackedMethod(invocation, c.SemanticModel);
+                    if (operands != null &&
+                        EquivalenceChecker.AreEquivalent(operands.Left, operands.Right))
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, operands.Left.GetLocation(),
+                            additionalLocations: new[] { operands.Right.GetLocation() },
+                            messageArgs: new[] { operands.Right.ToString(), operands.ErrorMessage }));
+                    }
+                }, SyntaxKind.InvocationExpression);
```

Comment:
Nope, I checked and the common pattern is to put it on new line. But this is so minor it's not worth arguing.